### DISCUSSION TITLE
Attribute access leaves transaction open / debugging tips

### DISCUSSION THREAD

**Discussion by aathan** [sqlalchemy/sqlalchemy#7109] [Usage Questions]
SQLA 1.4.23

Multiple sessions refer to the same underlying database, in which objects with lazy='joined' relationships exist.
A main "reader" session is used for querying and via the sqlalchemy.engine logger can be seen to be emitting selects that are fetching the related objects.
Meanwhile in another greenlet, writes are happening via a different "writer" session on the same database.

Sometimes the write deadlocks on commit() because the joined load session didn't commit

What's not clear to me is whether gevent has interrupted SQLA's engine right before it would have done a commit, or whether it is a bug, or intentional, that the select session's transaction is left uncommitted after the query has completed.

It looks like the relevant code is in the call stack of strategies.py _emit_lazyload() but it's not immediately apparent whether there would eventually be a commit() before the _emit_lazyload() completes -- or if that's even desirable. However, if (in the writer greenlet) I reader.commit() the reader session, the deadlock is avoided, so this seems like a reasonable question.

Possibly gevent has interrupted the lazyload just prior to a commit() OR, is what I'm describing an indication of a possible problem with a possible missing commit()? I have autocommit=False on these sessions. Is that the actual problem?

Any hints on where I should look to figure out whether the reader has been interrupted in an awkward place inside SQLA, perhaps right before it would have committed on its own?

---

I've set `sqlalchemy.engine` loging level to INFO, but I'm wondering if there are additional tips/tricks that you all can suggest in order to get more insight into what sqlalchemy is doing during attribute access etc. Particularly, I'd like to nab the line in the client code which is causing sqlalchemy to emit queries and/or open/close transactions -- I guess I can do that by trapping things at the logging formatter, but I'm wondering what a best practice for this would be. E.g., I don't want to just know what SQL is being emitted, but also the stack trace at the point where it is being emitted ... at least for certain classes or attributes or objects (yes, I know the answer to "how would that be specified" implies I'd need to hack a facility together).

Anyway ... ultimately what I'm trying to figure out is why SQLA is leaving a transaction open, in the DB (postgres) which appears to be from a select/join used to fetch a one to many relationship container's contents. The open transaction then deadlocks with other operations. From what I can tell, processing of the select's results should have completed, but the transaction remains open.

Yes, this is in a greenlet (user space multi-processing) context, where I collect objects that need to be persisted and then move them en-masse into a separate SQLA session to write them, then put them back into their original session. The deadlock happens between the "main" session, where the implicit fetch from the attribute access should be complete and the transaction comitted, and the new writer session, which can't finish the insert because of that other uncommitted session's select.

I can break the deadlock by issuing a commit() to the "main" session, but that doesn't seem safe.

**Comment by zzzeek** [comment]
I apologize this hasn't been answered as I see this referred towards in your newer question #7126.   gevent code can be more difficult to debug as it has a more unusual scheduling model going on, but overall if you have everything monkeypatched (e.g. threading.Lock, etc) it should act more or less like a threaded application.

the short answer, in absense of any specifics, is that any Session starts up a database transaction when it is first asked to emit SQL to a database.  That transaction stays open until one of rollback(), commit(), or close() is called.  

Where things usually become confusing is that objects which are still attached to this Session, which likely had its transaction committed already, are asked to return information about unloaded attributes.  When this occurs, assuming they are still attached to that session, they emit SQL, which means the owning Session starts a new transaction.

for this reason it is very advisable to make sure the Session stays around in the code until all objects related to it are discarded; or that session.close() is called to explicitly de-associate those objects from the session if database tranasctional work is done.    There is also the [raiseload](https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html#prevent-lazy-with-raiseload) strategy that's used to disallow relationship lazy loads of any kind unless explicitly allowed.
